src/game/card.rs

```diff
@@ -0,0 +1,56 @@
+#![allow(dead_code)]
+use rand::{seq::SliceRandom, thread_rng};
+use std::ops::{Deref, DerefMut};
+
+pub struct Monad;
+
+#[derive(Clone, Copy)]
+pub enum Color {
+    Red,
+    Orange,
+    Yellow,
+    Purple,
+    Blue,
+    Green,
+}
+
+pub enum Value {
+    Common,
+    Bi,
+    Tri,
+    Quad,
+    Quint,
+}
+
+pub struct Card(pub Value, pub Color);
+
+#[derive(Default)]
+pub struct Deck(Vec<Card>);
+
+impl Deck {
+    pub fn multiple(multiple: usize) -> Self {
+        Deck(Vec::with_capacity(COLORS.len() * multiple))
+    }
+
+    pub fn shuffle(&mut self) {
+        self.0.shuffle(&mut thread_rng());
+    }
+}
+
+impl Deref for Deck {
+    type Target = Vec<Card>;
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl DerefMut for Deck {
+    fn deref_mut(&mut self) -> &mut <Self as Deref>::Target { &mut self.0 }
+}
```

Comment:
Why are these two implementations important?

Reply:
These allow you to dereference a Deck into its Vec; dereferencing happens implicitly in method calls, which is why you can call `str`'s methods on a `String`. This lets you automatically call `Vec`'s methods on a `Deck`.

Deref is for turning `&Deck` into `&Vec<Card>`, and DerefMut is for `&mut Deck` into `&mut Vec`. These are both important because they correspond to methods like `fn foo(&self)` and `fn foo(&mut self)` respectively.

Reply:
How does `DerefMut` have `Target` in scope? `Target` is declared within `Deref`. Is it because `<Self as Deref>` brings it into scope because of the type change to `Deref`? Also, what does `<thing>` mean and how is it used?

Reply:
`<Struct as Trait>` or `<object as Trait>` is a form of casting; in particular, it's used for disambiguation. By statically casting the object to a Trait Object, we can [disambiguate method calls](https://doc.rust-lang.org/book/second-edition/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name) that may have namespace collisions in different traits, and also access a specific trait's [Associated Types](https://doc.rust-lang.org/book/second-edition/ch19-03-advanced-traits.html#specifying-placeholder-types-in-trait-definitions-with-associated-types). This is very useful in generic trait definitions, and sometimes is used in concrete implementations like this.

`<Self as Deref>` accesses exclusively the `Deref` implementation for `Self` (here `Deck`). Since `Deref::Target` is a public associated type, we can reuse it wherever we like and only have to change it in one place in the future.